### ISSUE TITLE
v1.32.0 results in lots of i/o timeouts

### ISSUE THREAD

**Issue by deepakprabhakara** [aerospike/aerospike-client-go#227]
I updated from v1.31.0 to v1.32.0 and notice a lot of "read tcp 10.20.105.2:47888->10.20.73.2:3000: i/o timeout" errors. Are there any configuration changes needed to migrate from v1.31.0 to v1.32.0?

**Comment by khaf** [comment]
Thanks for your report. I'm not aware of any, but it seems other users are experiencing similar issue.

While I'm not sure is the issues are related, I'm investigating and will issue a fix ASAP.

**Comment by deepakprabhakara** [comment]
Thanks @khaf. I suspect maybe stale connections might be left behind and only surfaces during high traffic.

**Comment by clusc** [comment]
Not specifically related to v1.32.0, but we've been seen i/o timeouts consistently too. It seems related to the concurrency level while using the go library. If we limit the concurrency(by using a semaphore for example), we could achieve a lot higher throughputs and lower io timeout issues. The concurrency level is very case dependent, so not sure we want to put it into the client library directly.

**Comment by khaf** [comment]
It seems that a change in the default value for the `Policy.SocketTimeout` from 0 to 100ms is the culprit here.

Can you check if you are setting your own appropriate `Timeout` / `SocketTimeout` in your policies? If not, I have to dig much deeper. This issue is only reproducible for me with that timeout change.

**Comment by khaf** [comment]
@clusc What you are seeing is having too many requests to a machine and bogging Go's epoll queue.

By using a lock, you are using a crude implementation of a Bucket rate limiter. A more practical way would be to rate-limit your app on the load-balancer or the reverse-proxy.

**Comment by deepakprabhakara** [comment]
@khaf Aah yes, default value of 'Policy.SocketTimeout' could be the issue. I'll test it with 0 and let you know.

**Comment by fmstephe** [comment]
@clusc which version are you currently running?

**Comment by clusc** [comment]
@fmstephe we are using v1.32.0, but have our own timeout sets always.
@khaf Rate limiting the app is not really feasible in our use case. We do need to serve that many concurrent request, and it's not that aerospike can't handle it, it's just the client aren't interacting with the network stack efficient enough. What is practical possible is to scale down and up our client application, effectively creating more epoll queue to serve the same amount of traffic. We can see what we can do in this direction.

**Comment by khaf** [comment]
Does anyone have any news regarding the issue? Did the suggested changes make a difference?

**Comment by khaf** [comment]
This issue may be related to the change in the newer versions in which write operations ONLY go to the master node, and the `getMasterNode` returns an error instead of a random node if the partition map is empty or in flux. Take a look at [this](https://github.com/aerospike/aerospike-client-go/issues/229) ticket to see if it is related.

**Comment by khaf** [comment]
Another issue is that the BatchPolicy in the older release was not initializing the default policy it embeds properly. That issue has been taken care of and already in master, and may improve the batch transaction timeouts in case you were using the default values (which is not recommended).

**Comment by deepakprabhakara** [comment]
@khaf Sorry never got a chance to test this again, I ended up pinning the previous version of the lib. Will post back here when I update next.

**Comment by cgebe** [comment]
I am running into this issue on 1.32.1, i have one client which is used by ~300 goroutines. Each goroutine pushes ~2000 data points at startup, which causes i/o timeouts to show up in the log:

```
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:37312-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:36164-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:41464-&gt;10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:36824-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:37582-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:36194-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:39936-&gt;10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:37366-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:39212-&gt;10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:57140-&gt;10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:35378-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:37132-&gt;10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:57582-&gt;10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538227171,"message":"worker: insert pack read tcp 10.244.1.102:37170-&gt;10.244.6.233:3000: i/o timeout"} 
```

Which parameters sets the socket timeout in the latest client? All available CientPolicy params seem to be used for other stuff.

**Comment by khaf** [comment]
The actual socket timeout is `XXXPolicy.SocketTimeout`. I'm very interested in resolving this issue, but I need clarification about the circumstances in which it happens. From the looks of it, it is use-case dependent, which is difficult for me to reproduce and test.

**Comment by cgebe** [comment]
I just tried it with creating a dedicated client for each goroutine, same result - i/o errors. 
Setting the `SocketTimeout` to 0 does not produce errors anymore, obviously. However setting it to 1000, still errors showing up. By setting it to 1000, additionally `maxRetries` errors are reported:

```
{"level":"info","time":1538228858,"message":"worker: insert pack read tcp 10.244.6.5:40626-&gt;10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538228858,"message":"worker: insert pack command execution timed out on client: Exceeded number of retries. See `Policy.MaxRetries`. (last error: write tcp 10.244.6.5:40726-&gt;10.244.7.42:3000: i/o timeout)"}
{"level":"info","time":1538228858,"message":"worker: insert pack command execution timed out on client: Exceeded number of retries. See `Policy.MaxRetries`. (last error: write tcp 10.244.6.5:41046-&gt;10.244.7.42:3000: i/o timeout)"}
{"level":"info","time":1538228858,"message":"worker: insert pack command execution timed out on client: Exceeded number of retries. See `Policy.MaxRetries`. (last error: write tcp 10.244.6.5:40808-&gt;10.244.7.42:3000: i/o timeout)"}
{"level":"info","time":1538228858,"message":"worker: insert pack read tcp 10.244.6.5:57942-&gt;10.244.3.82:3000: i/o timeout"} 
```

I am running aerospike in a 4 node cluster with 20GB memory on each node. The application pushing data runs on the same nodes. The nodes are located in the same data center, so no latency issues. I keep investigating. The throughput seems much to low to cause this performance issue.

**Comment by cgebe** [comment]
Errors always coming in 'batches', see report times:


<details><summary>Log</summary>
<p>

```
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:59662->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37502->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37178->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38234->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:57292->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36568->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37148->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35918->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35898->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:34658->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38876->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:34462->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:57004->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38158->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:34484->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:34016->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38628->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36484->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:57574->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38240->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36324->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35632->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37794->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36608->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37846->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36412->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:57544->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36512->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:58160->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:58366->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35534->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37648->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38236->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37582->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:59354->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35636->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35794->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:40364->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:60606->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:59698->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35482->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:56766->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37058->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:36364->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:59666->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:59392->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:38222->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:60990->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:37372->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230063,"message":"worker: insert pack read tcp 10.244.1.105:35462->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:59912->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:36342->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:59284->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:36250->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:38098->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:35490->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:38374->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:36768->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:36272->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:38296->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:39236->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:34062->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:34202->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:60852->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:36072->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:34184->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:37290->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:35960->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:37534->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:60148->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:37238->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:37744->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:38864->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:59910->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:59512->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:34652->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:60206->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:35170->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:60310->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:37180->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:34364->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:38206->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230181,"message":"worker: insert pack read tcp 10.244.1.105:39772->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:33774->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:38720->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:34502->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:33590->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36882->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37996->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:35368->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36218->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:38246->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:58512->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36610->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:60128->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37842->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:34336->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36402->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:38672->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37836->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:57140->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:38318->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:59116->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36348->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36010->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:58550->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37966->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:58944->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:35500->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:36906->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37582->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37886->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37760->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37644->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:40198->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37348->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:33560->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:57996->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:34474->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230238,"message":"worker: insert pack read tcp 10.244.1.105:37742->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:59950->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:39900->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:60274->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37130->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:34646->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:36934->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:36372->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:36944->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:59460->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:60164->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:39064->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37262->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37152->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:38446->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37416->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37880->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:33700->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:60982->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:56712->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37522->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:58112->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:35846->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:37114->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:38860->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:35522->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:36070->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:59980->10.244.7.42:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:35122->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:38088->10.244.1.77:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:36796->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:35246->10.244.6.233:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:32856->10.244.3.82:3000: i/o timeout"}
{"level":"info","time":1538230262,"message":"worker: insert pack read tcp 10.244.1.105:38666->10.244.1.77:3000: i/o timeout"}
```

</p>
</details>

**Comment by khaf** [comment]
Are these batch requests or simple write/reads? And can you set the `SocketTimeout` to 0 and measure the latency yourself on the application side to see how long is the maximum?

One final thought is that the initial timeouts may be related to the length of time it takes to connect and fill up the connection pool, and after that things may perform better.

**Comment by cgebe** [comment]
I measured the times. Right at the start of the application, the max increases:
Startup on `1538232837`
```
{"level":"info","time":"1538232837","message":"new max 10.931254ms"}
{"level":"info","time":"1538232837","message":"new max 35.127487ms"}
{"level":"info","time":"1538232837","message":"new max 38.892417ms"}
{"level":"info","time":"1538232837","message":"new max 72.38891ms"}
{"level":"info","time":"1538232837","message":"new max 79.287429ms"}
{"level":"info","time":"1538232837","message":"new max 109.961144ms"} 
{"level":"info","time":"1538225968","message":"new max 268.362566ms"}
{"level":"info","time":"1538225968","message":"new max 283.688369ms"}
{"level":"info","time":"1538226029","message":"new max 327.32074ms"}
{"level":"info","time":"1538226029","message":"new max 330.441434ms"} 
```

> One final thought is that the initial timeouts may be related to the length of time it takes to connect and fill up the connection pool, and after that things may perform better.

Indeed. The errors happen periodically, i guess as soon as the connection pool is revamped. Did not try https://github.com/aerospike/aerospike-client-go/commit/6b0509afa616dde0d210f0683721dd866ce8e3e3 yet.

**Comment by khaf** [comment]
Interesting. To make sure we're on the write track, try to run a lot of goroutines with simple reads to warm up and fill the connection queues in the beginning just before starting the writes. Let's see if that helps with the timeouts.

**Comment by cgebe** [comment]
The current program flow already reads out before each write.

1. Read data from aerospike
2. Process data to new representation
3. Write data to aerospike

**Comment by cgebe** [comment]
The issue occurs when reading only as well. I adjusted my code in order to check for the existence of an object before fetching data, converting it and writing it back.
